### PR TITLE
core: implement displayState for Stage in avm1/2

### PR DESCRIPTION
**Pull request by hatal175**
Relevant to several issues (#1562, #4785, #274).

This pull request implements the displayState member for Stage for both avm1 and avm2. I had to rewrite several parts so fullscreen actions would go through the Stage class in core.
I've also implemented the full screen events for both AVMs.

I would say that the non-interactive part of full screen is not implemented.

Several issues that I've run into are the following:
1. weird rendering issues in log in desktop about extent. It seemed to still go to full screen, so yay.
2. in web, the fullscreen functions are asynchronous. This was happily ignored before and still is but it could cause issues.
3. the state can be fullScreenInteractive even though you set it to fullScreen. No clue what's that about.

### CODE REVIEW

**Comment by hatal175** [comment]
Not sure what's up with that check. Doesn't seem related to change.

**Comment by torokati44** [comment]
I have seen that same check spuriously fail in a different PR as well.
The error message `element ("<ruffle-object />") still not existing after 30000ms` hints that something could have hanged, or possibly there is a race condition?

**Comment by kmeisthax** [comment]
The error you're getting is a recurring problem with our CI infrastructure. Sometimes the web tests just fail to create the Ruffle element correctly, and we haven't been able to determine why. The solution in every case is to just force the test to run again.

**Comment by adrian17** [comment]
A general thing I just noticed in docs:

> Full-screen mode is initiated in response to a mouse click or key press by the user; the movie cannot change Stage.displayState without user input

> While Flash Player is in full-screen mode, all keyboard input is disabled (except keyboard shortcuts that take the user out of full-screen mode). 

I imagine the latter isn't really relevant anymore now that we use "browser fullscreen", but the former sounds reasonable - we probably don't want movies to fullscreen themselves at will? Or do browsers handle that automatically even if there's wasm in the middle?

I'm not saying it's a blocker or obligatory to implement, but that does sound like something we should maybe talk about here or on discord?

As for AVM2 stuff, probably best for kmeisthax to review this.

**Comment by hatal175** [comment]
Only within an html page and that's actually enforced by browser.

File: core/src/player.rs [hunk 1]
```diff
@@ -677,6 +677,17 @@ impl Player {
         );
     }
 
+    pub fn run_context_menu_fullscreen_callback(&mut self, is_fullscreen: bool) {
```
Comment (kmeisthax) [inline core/src/player.rs:680]: This should probably be called `set_is_fullscreen`, or `did_enter_fullscreen`, or something. This name is a bit too verbose and less descriptive as to what it does.

**Review by kmeisthax — CHANGES_REQUESTED** [review]
I have one very minor nitpick, and it's not even with the AVM2 stuff. Everything else seemed straightforward.

**Comment by danielhjacobs** [comment]
I don't know if you meant to (I see a force push and a review request so I assume you did) but you never changed the name of the method run_context_menu_fullscreen_callback despite marking the conversation as resolved.

**Comment by hatal175** [comment]
I might have misunderstood. I thought he talked about the parameter name.
That said, it's pretty much like another context function...

**Comment by danielhjacobs** [comment]
Yeah, it's consistent with the names @adrian17 gave the functions added in #4173. I can't say I like those names, but it does match what already exists, so perhaps it should be kept. That would be @kmeisthax 's call though.

**Comment by adrian17** [comment]
> That said, it's pretty much like another context function...

> Yeah, it's consistent with the names @adrian17 gave the functions added in #4173

Not really. The intention behind `run_context_menu_callback` is to be a single point from context menu item to a callback function attached to it. IMO if we wanted to be consistent with that interface, instead of adding `run_context_menu_fullscreen_callback`, the Fullscreen button should be handled the same way as all other builtin context menu items, and there should be another branch added to the former function, like this:

```rs
ContextMenuCallback::Play => Self::toggle_play_root_movie(context),
ContextMenuCallback::Forward => Self::forward_root_movie(context),
ContextMenuCallback::Back => Self::back_root_movie(context),
ContextMenuCallback::Rewind => Self::rewind_root_movie(context),
ContextMenuCallback::Fullscreen=> Self::toggle_fullscreen(context), // or Enable/DisableFullscreen pair?
```

**Comment by hatal175** [comment]
Renamed function.

**Comment by danielhjacobs** [comment]
> > IMO if we wanted to be consistent with that interface, instead of adding `run_context_menu_fullscreen_callback`, the Fullscreen button should be handled the same way as all other builtin context menu items, and there should be another branch added to the former function

Then it would be hidden by menu="0", and I disagree with doing that personally, as I mentioned here:https://github.com/ruffle-rs/ruffle/issues/4258#issuecomment-864100254

**Comment by adrian17** [comment]
IMO, using a single API internally doesn't mean that internal logic can't be changed to always show `Fullscreen`.

Anyway, this can still be done in a separate PR if needed; as for now, the `set_fullscreen` is a nice rename, thanks :)

**Review by kmeisthax — APPROVED** [review]

**Comment by kmeisthax** [comment]
@hatal175 Sorry for the wait, I'll merge this on the next rebase

**Comment by ousia** [comment]
Many thanks for the implementation, @hatal175.

**Pull request merged**